### PR TITLE
Bakerstu dcc accessory client

### PR DESCRIPTION
**Pull request by bakerstu**
I'm not ready to merge this back, but I would like you to take a quick look to see if I'm on the right track at least.  I'm interested in your feedback.  My next step is to write some unit tests.

### CODE REVIEW

File: src/openlcb/NonAuthoritativeEventProducer.cxxtest [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+#include "NonAuthoritativeEventProducer.hxx"
+
+using ::testing::_;
+using ::testing::Mock;
+using ::testing::StrictMock;
+
+namespace openlcb
+{
+
+static const uint64_t kEventBase = 0x05010101FFFF0000ULL;
```
Comment (balazsracz) [inline src/openlcb/NonAuthoritativeEventProducer.cxxtest]: 0501010118FF0000
(just so that it is in our range)

File: src/openlcb/NonAuthoritativeEventProducer.cxxtest [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    // dual range
+    expect_packet(":X1952422AN05010101FFFF004F;");
+    expect_packet(":X1952422AN05010101FFFF006F;");
+
+    send_packet(":X19970001N;");
+    wait_for_event_thread(); Mock::VerifyAndClear(&canBus_);
+}
+
+TEST_F(BitRangeNonAuthoritativeEventPTest, GlobalIdentifyAddressed)
```
Comment (balazsracz) [inline src/openlcb/NonAuthoritativeEventProducer.cxxtest]: alkso add a negative example, when no packets are expected.

File: src/openlcb/NonAuthoritativeEventProducer.cxxtest [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+    // dual range
+    expect_packet(":X1952422AN05010101FFFF004F;");
+    expect_packet(":X1952422AN05010101FFFF006F;");
+
+    send_packet(":X19970001N022A;");
+    wait_for_event_thread(); Mock::VerifyAndClear(&canBus_);
+}
+
+TEST_F(BitRangeNonAuthoritativeEventPTest, IdentifyProducer)
```
Comment (balazsracz) [inline src/openlcb/NonAuthoritativeEventProducer.cxxtest]: do you have any tests for queries?

File: src/os/MDNS.cxx [hunk 4]
```diff
@@ -41,6 +41,7 @@
 void mdns_publish(const char *name, const char *service, uint16_t port) __attribute__ ((weak));
 int mdns_lookup(const char *service, struct addrinfo *hints,
                 struct addrinfo **addr) __attribute__ ((weak));
+void mdns_scan(const char *service) __attribute__ ((weak));
```
Comment (balazsracz) [inline src/os/MDNS.cxx]: these changes probably belong to a different PR

File: src/openlcb/DccAccyProducer.hxx [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+public:
+    /// DCC accessory activation values
+    enum
+    {
+        NORMAL = false, ///< normal route
+        REVERSE = true, ///< reverse route
+    };
+
+    /// highest possible DCC address supported
+    static constexpr uint16_t MAX_ADDRESS = 2040;
```
Comment (bakerstu) [inline src/openlcb/DccAccyProducer.hxx]: I've contemplated that as well.  Not sure if we want to support broadcast, or perhaps even make it an option.

File: src/openlcb/NonAuthoritativeEventProducer.cxx [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+            range = EncodeRange(eventBase_,
+                                eventBaseOff_ == 0 ? size_ * 2 : size_);
+        }
+        else if (entry.event == eventBaseOff_ && eventBaseOff_ != 0)
+        {
+            range = EncodeRange(eventBaseOff_, size_);
+        }
+        else
+        {
+            // uninteresting event range
```
Comment (bakerstu) [inline src/openlcb/NonAuthoritativeEventProducer.cxx]: Is that true even if my event range is not an exact power of 2?

File: src/openlcb/NonAuthoritativeEventProducer.cxxtest (hunk 2 above)
Comment (bakerstu) [inline src/openlcb/NonAuthoritativeEventProducer.cxxtest]: Yeah, I wanted to do this.  Howe do I test for that?

File: src/openlcb/NonAuthoritativeEventProducer.hxx [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+ */
+
+#ifndef _OPENLCB_NONAUTHORITATIVEEVENTPRODUCER_HXX_
+#define _OPENLCB_NONAUTHORITATIVEEVENTPRODUCER_HXX_
+
+#include "openlcb/EventHandlerTemplates.hxx"
+
+namespace openlcb
+{
+#if 0
```
Comment (bakerstu) [inline src/openlcb/NonAuthoritativeEventProducer.hxx]: That is the intent.  It was copied here as a quick reference.

File: src/openlcb/NonAuthoritativeEventProducer.hxx [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+public:
+    /// Constructor.  Creates a new bit range producer.
+    ///
+    /// @param node the node that the producer will be bound to
+    /// @param event_base the starting event ID for the event range
+    /// @param size The total event range size in "pairs" of events.  Each
+    ///             pair of sequential events represents a single "bit" with
+    ///             a binary state.
+    /// @param state_callback Callback method for delivering the results of a
+    ///                       consumer identified.  The first unsigned parameter
```
Comment (bakerstu) [inline src/openlcb/NonAuthoritativeEventProducer.hxx]: The code already calls the same callback for both "event report" and "consumer identified".  The comment should be updated accordingly.

File: src/openlcb/NonAuthoritativeEventProducer.hxx [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    /// Handle an incoming consumer identified message.
+    ///
+    /// @param entry reference to this entry in the event registry
+    /// @param event event metadata
+    /// @param done notifible to wakup when finished
+    void handle_consumer_identified(const EventRegistryEntry &entry,
+                                    EventReport *event,
+                                    BarrierNotifiable *done) override;
+
+    /// Handle an incoming identify producer message.
```
Comment (bakerstu) [inline src/openlcb/NonAuthoritativeEventProducer.hxx]: fixed

Comment (bakerstu) [inline src/openlcb/NonAuthoritativeEventProducer.hxx]: Actually, is it not both?  Which hander gets called for an Identify Events Addressed?

File: src/os/MDNS.cxx (hunk 4 above)
Comment (bakerstu) [inline src/os/MDNS.cxx]: I thought I put those on another branch.  I'll have to look into this further.

File: src/openlcb/NonAuthoritativeEventProducer.cxx (hunk 6 above)
Comment (balazsracz) [inline src/openlcb/NonAuthoritativeEventProducer.cxx]: Actually you're right, there is a problem when the base is not aligned. The big issue is that entry.event == eventBase_ will not be true at all in that case. The registered event is already rounded up to the power of two boundary, while eventBase is not.

If we do the encoding of which eventbase we used in the userdata uint32_t in the Entry, then we can use that to figure out which one to report here.

Comment (balazsracz) [inline src/openlcb/NonAuthoritativeEventProducer.cxx]: Also we need to add tests for cases when the base is not aligned.

File: src/openlcb/NonAuthoritativeEventProducer.cxxtest (hunk 2 above)
Comment (balazsracz) [inline src/openlcb/NonAuthoritativeEventProducer.cxxtest]: send_packet(":X19970001N022B;");
wait_for_event_thread(); Mock::VerifyAndClear(&canBus_);

File: src/openlcb/NonAuthoritativeEventProducer.hxx (hunk 9 above)
Comment (balazsracz) [inline src/openlcb/NonAuthoritativeEventProducer.hxx]: yes it is both. Hence the documentation should say "identify global and addressed" [/ global or addressed?], sorry for not being clear.

File: src/openlcb/NonAuthoritativeEventProducer.cxx (hunk 6 above)
Comment (bakerstu) [inline src/openlcb/NonAuthoritativeEventProducer.cxx]: Is this not also a problem on the top end as well?  For example, if the range ends before a power of 2 boundary, I should not respond to a request beyond the end of my range.
